File: 백준/Silver/2583. 영역 구하기/영역 구하기.js
const input = require('fs')
  .readFileSync(0)
  .toString()
  .trim()
  .split('\n');
const [M, N, K] = input.shift().split(' ').map(Number);

const graph = Array.from(Array(M), () => Array(N).fill(0));
const answer = [];

// 입력값을 그래프로 변환하기 위한 반복문
for (let i = 0; i < K; i += 1) {
  const [x1, y1, x2, y2] = input[i].split(' ').map(Number);

  // 직사각형 영역은 1로 표시, 빈 영역은 0으로 표시
  for (let y = M - y2; y < M - y1; y += 1) {
    for (let x = x1; x < x2; x += 1) {
      graph[y][x] = 1;
    }
  }
}

function bfs(start) {
  const ds = [
    [-1, 0],
    [1, 0],
    [0, 1],
    [0, -1],
  ]; // 좌우상하
  const queue = [start]; // 시작점은 무조건 돌아야 하는 노드이기 때문에 큐에 넣고 시작

  let count = 0;

  while (queue.length) {
    // 큐에 남아 있는 노드가 없다면 탐색이 종료된 것
    const [cy, cx] = queue.shift();
    count += 1;

    for (let i = 0; i < 4; i += 1) {
      const ny = cy + ds[i][1];
      const nx = cx + ds[i][0];

      // 해당 위치가 그래프를 벗어나지 않고, 아직 방문하지 않은 노드(0)라면 방문 처리 후 큐에 담기
      if (ny >= 0 && ny < M && nx >= 0 && nx < N && !graph[ny][nx]) {
        graph[ny][nx] = 1;
        queue.push([ny, nx]);
      }
    }
  }

  return count;
}

for (let i = 0; i < M; i += 1) {
  for (let j = 0; j < N; j += 1) {
    // 아직 방문하지 않은 노드(0)라면
    if (!graph[i][j]) {
      graph[i][j] = 1; // 방문 처리
      answer.push(bfs([i, j])); // bfs를 실행해 영역 칸 개수를 배열에 담음
    }
  }
}

console.log(answer.length);
console.log(answer.sort((a, b) => a - b).join(' '));

File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
from collections import deque

def bfs(x, y):
    count = 0
    queue = deque([(x, y)])
    visited[x][y] = 1

    while queue:
        cx, cy = queue.popleft()
        count += 1
        for dx, dy in dir:
            nx = cx + dx
            ny = cy + dy

            if 0 <= nx < M and 0 <= ny < N and not paper[nx][ny] and not visited[nx][ny]:
                visited[nx][ny] = 1
                queue.append((nx, ny))

    sizes.append(count)

M, N, K = map(int, input().split())
paper = [[0] * N for _ in range(M)]
visited = [[0] * N for _ in range(M)]

dir = [[-1, 0], [1, 0], [0, -1], [0, 1]]
sizes = []

for _ in range(K):
    x1, y1, x2, y2 = map(int, input().split())
    for i in range(x1, x2):
        for j in range(y1, y2):
            paper[j][i] = 1

count = 0
for i in range(M):
    for j in range(N):
        if not paper[i][j] and not visited[i][j]:
            bfs(i, j)
            count += 1

print(count)
print(' '.join(map(str, sorted(sizes))))